File: todo/todo.go
package todo

import (
	"encoding/json"
	"log"
	"net/http"
	"strconv"
	"sync"
)

type Todo struct {
	ID   int    `json:"id"`
	Text string `json:"text"`
	Done bool   `json:"done"`
}

var (
	todos = []Todo{
		{ID: 1, Text: "Learn Eronom", Done: false},
		{ID: 2, Text: "Add todo API in Gosssssvs gooooo", Done: true},
	}
	nextID = 3
	mu     sync.Mutex
)

// Helper to write JSON responses
func WriteJSON(w http.ResponseWriter, status int, data interface{}) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	if err := json.NewEncoder(w).Encode(data); err != nil {
		log.Printf("Error encoding JSON: %v", err)
	}
}

func HandleGetTodos(w http.ResponseWriter, r *http.Request) {
	mu.Lock()
	defer mu.Unlock()
	WriteJSON(w, http.StatusOK, todos)
}

func HandleCreateTodo(w http.ResponseWriter, r *http.Request) {
	var input struct {
		Text string `json:"text"`
	}
	if err := json.NewDecoder(r.Body).Decode(&input); err != nil {
		WriteJSON(w, http.StatusBadRequest, map[string]string{"error": "Invalid input"})
		return
	}
	if input.Text == "" {
		WriteJSON(w, http.StatusBadRequest, map[string]string{"error": "Text is required"})
		return
	}

	mu.Lock()
	newTodo := Todo{
		ID:   nextID,
		Text: input.Text,
		Done: false,
	}
	nextID++
	todos = append(todos, newTodo)
	mu.Unlock()

	WriteJSON(w, http.StatusCreated, newTodo)
}

func HandleUpdateTodo(w http.ResponseWriter, r *http.Request) {
	idStr := r.PathValue("id")
	id, err := strconv.Atoi(idStr)
	if err != nil {
		WriteJSON(w, http.StatusBadRequest, map[string]string{"error": "Invalid ID"})
		return
	}

	var input struct {
		Text *string `json:"text"`
		Done *bool   `json:"done"`
	}
	if err := json.NewDecoder(r.Body).Decode(&input); err != nil {
		WriteJSON(w, http.StatusBadRequest, map[string]string{"error": "Invalid input"})
		return
	}

	mu.Lock()
	defer mu.Unlock()

	for i, todo := range todos {
		if todo.ID == id {
			if input.Text != nil {
				todos[i].Text = *input.Text
			}
			if input.Done != nil {
				todos[i].Done = *input.Done
			}
			WriteJSON(w, http.StatusOK, todos[i])
			return
		}
	}

	WriteJSON(w, http.StatusNotFound, map[string]string{"error": "Todo not found"})
}

func HandleDeleteTodo(w http.ResponseWriter, r *http.Request) {
	idStr := r.PathValue("id")
	id, err := strconv.Atoi(idStr)
	if err != nil {
		WriteJSON(w, http.StatusBadRequest, map[string]string{"error": "Invalid ID"})
		return
	}

	mu.Lock()
	defer mu.Unlock()

	for i, todo := range todos {
		if todo.ID == id {
			todos = append(todos[:i], todos[i+1:]...)
			WriteJSON(w, http.StatusOK, map[string]string{"status": "deleted"})
			return
		}
	}

	WriteJSON(w, http.StatusNotFound, map[string]string{"error": "Todo not found"})
}

File: todo/todo.py
from fastapi import APIRouter, Request, status
from fastapi.responses import JSONResponse
import json

router = APIRouter()

# Shared in-memory list of todos
todos = [
    {"id": 1, "text": "Learn Eronom", "done": False},
    {"id": 2, "text": "Add todo API in pyzzz", "done": True}
]
next_id = 3

@router.get("/todos")
async def get_todos():
    """Retrieve all todos."""
    return todos

@router.post("/todos")
async def create_todo(request: Request):
    """Create a new todo."""
    try:
        body = await request.body()
        input_data = json.loads(body.decode('utf-8'))
    except Exception:
        return JSONResponse(
            status_code=status.HTTP_400_BAD_REQUEST,
            content={"error": "Invalid input"}
        )

    text = input_data.get("text")
    if not isinstance(text, str) or text == "":
        return JSONResponse(
            status_code=status.HTTP_400_BAD_REQUEST,
            content={"error": "Text is required"}
        )

    global next_id
    new_todo = {
        "id": next_id,
        "text": text,
        "done": False
    }
    next_id += 1
    todos.append(new_todo)
    return JSONResponse(status_code=status.HTTP_201_CREATED, content=new_todo)

@router.put("/todos/{id}")
async def update_todo(id: str, request: Request):
    """Update an existing todo by ID."""
    try:
        todo_id = int(id)
    except ValueError:
        return JSONResponse(
            status_code=status.HTTP_400_BAD_REQUEST,
            content={"error": "Invalid ID"}
        )

    try:
        body = await request.body()
        input_data = json.loads(body.decode('utf-8'))
    except Exception:
        return JSONResponse(
            status_code=status.HTTP_400_BAD_REQUEST,
            content={"error": "Invalid input"}
        )

    for item in todos:
        if item["id"] == todo_id:
            if "text" in input_data and input_data["text"] is not None:
                item["text"] = str(input_data["text"])
            if "done" in input_data and input_data["done"] is not None:
                item["done"] = bool(input_data["done"])
            return item

    return JSONResponse(
        status_code=status.HTTP_404_NOT_FOUND,
        content={"error": "Todo not found"}
    )

@router.delete("/todos/{id}")
async def delete_todo(id: str):
    """Delete a todo by ID."""
    try:
        todo_id = int(id)
    except ValueError:
        return JSONResponse(
            status_code=status.HTTP_400_BAD_REQUEST,
            content={"error": "Invalid ID"}
        )

    for i, item in enumerate(todos):
        if item["id"] == todo_id:
            todos.pop(i)
            return {"status": "deleted"}

    return JSONResponse(
        status_code=status.HTTP_404_NOT_FOUND,
        content={"error": "Todo not found"}
    )
